### PR TITLE
Update i18nTextCollector to collect strings also from themes

### LINKED ISSUE #8491: i18nTextCollector not collecting from themes
## Description

It's currently not possible to run something like `/dev/tasks/i18nTextCollectorTask/?module=themes/mytheme` to let I18NTextcollector collect entities from my theme.

Afaik this was possible in SS3.

## PRs
- https://github.com/silverstripe/silverstripe-framework/pull/10685
- https://github.com/silverstripe/developer-docs/pull/263

### PR DESCRIPTION
**Pull request by xini**
Fixes #8491

### CODE REVIEW

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (michalkleiner) [inline src/i18n/TextCollection/i18nTextCollector.php]: I'm not 100% convinced about the method name since it doesn't return just standard modules, but that's probably ok as long as we change the visibility to `private` and add a docblock describing that it includes themes in the list as well.

**Review by michalkleiner — CHANGES_REQUESTED** [review]

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php]: Thank you very much. I have fixed that in the latest commit.

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (michalkleiner) [inline src/i18n/TextCollection/i18nTextCollector.php]: Sorry, didn't notice this before. We can't use `str_starts_with` as that is PHP 8 only, we still need to support PHP 7.4 in 4.x framework, so please use `strpos`.

**Review by michalkleiner — CHANGES_REQUESTED** [review]

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php]: done

**Review by michalkleiner — DISMISSED** [review]
Looks ok from my perspective, thanks for the tweaks!

I'm not too fussed if this particular PR goes to `4.12` as a bugfix or to `4` as an enhancement, what do you think @GuySartorelli?

P.S. Squash will be needed

**Comment by michalkleiner** [comment]
> I'm not too fussed if this particular PR goes to `4.12` as a bugfix or to `4` as an enhancement, what do you think @GuySartorelli?

Based on Guy's [comment](https://github.com/silverstripe/silverstripe-framework/pull/10686#issuecomment-1427356114) on #10686 I suggest this also targets `4`. Then we can get both in as enhancements for 4.13 which is not too far away.

**Comment by xini** [comment]
rebased and squashed.

**Comment by xini** [comment]
the tests didn't trigger when I rebased this. can you please trigger them manually?

**Comment by GuySartorelli** [comment]
There's no way that I'm aware of to trigger tests manually on a PR. Try pushing to the branch again.

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (GuySartorelli) [inline src/i18n/TextCollection/i18nTextCollector.php]: Why not use `Path::join()`?

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (GuySartorelli) [inline src/i18n/TextCollection/i18nTextCollector.php]: I assume the `THEMES_DIR . '/' . $theme` part of this isn't intended to represent an actual file path at all, and is just a way to distinguish themes from modules in `getEntitiesByModule()`?
So to run text collector against a theme, you run `sake dev/tasks/i18nTextCollectorTask module=themes/my-theme-name`?

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (GuySartorelli) [inline src/i18n/TextCollection/i18nTextCollector.php]: No need for this change - the path returned by `$module->getPath()` has already been through `Path::normalise()` which in turn uses `Convert::slashes()` - long story short, at this point we only have unix-style file paths.
This change could actually _stop_ this check from working correctly.

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (GuySartorelli) [inline src/i18n/TextCollection/i18nTextCollector.php]: Here we're relying directly on the array keys in returned from `getModulesAndThemes()`, but in `getEntitiesByModule()` we only rely on array keys for themes - we should be consistent and either always use the array key, or only ever use the array key for themes.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php]: correct, yes.

Comment (GuySartorelli) [reply src/i18n/TextCollection/i18nTextCollector.php]: In that case instead of using the `THEMES_DIR` constant just use the hardcoded string "themes" to avoid confusion

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php]: I wanted to keep as much of the existing functionality intact and only rely on the keys where needed. I can't see the woods for the trees. How would you change this here?

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php]: done. thanks!

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php]: done. thanks!

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php]: done, thanks!

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php]: Hey @GuySartorelli what do you want me to do here to get this over the line? Cheers

**Comment by GuySartorelli** [comment]
Sorry, I haven't had time to look at this again - at the very least this will need to target the `5` branch and be rebased. Can you please do that? I'll try make time to look at this properly over the next couple of weeks.

**Comment by xini** [comment]
Thanks @GuySartorelli. This is now rebased and squashed.

File: src/i18n/Messages/YamlWriter.php
Comment (GuySartorelli) [inline src/i18n/Messages/YamlWriter.php]: use `Path::join()` instead.

An argument could be made that this is unrelated to the change this PR is for, and it should be done in a separate PR. If you want _this_ part of the PR to be in CMS 4.13 as a bug fix, you'll need to do it as a separate PR. Otherwise, just make the change I suggested above and I'll be okay with it.

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (GuySartorelli) [inline src/i18n/TextCollection/i18nTextCollector.php:470]: This new method is called _four times_ in this collector - it should probably store the result in a `private` property the first time, and then just return that value the other three times.

Alternatively this should set a property, and only ever be called once. Then just directly reference the property all the other times. If you use this approach, the method should probably be called something like `findModulesAndThemes()` instead.

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (GuySartorelli) [inline src/i18n/TextCollection/i18nTextCollector.php]: Been thinking a bit more about this - we should probably use some delimiter we _know_ won't be used in an actual module name. Colons aren't allowed in composer names so this protects against the unlikely edge case of a module called `themes/my-module`

```suggestion
                    $modules['themes:' . $theme] = new Module(Path::join(THEMES_PATH, $theme), BASE_PATH);
```

Probably also makes sense to set `'themes:'` as a private const on this class, since it's referenced a few times and this makes it super clear that those usages do have to explicitly be the same value.
```php
private const THEME_PREFIX = 'themes:';
```

```suggestion
                    $modules[self::THEME_PREFIX . $theme] = new Module(Path::join(THEMES_PATH, $theme), BASE_PATH);
```

File: src/i18n/Messages/YamlWriter.php
Comment (GuySartorelli) [reply src/i18n/Messages/YamlWriter.php]: If you're going to keep this change in this PR it probably also makes sense to change `DIRECTORY_SEPARATOR` usages from `i18nTextCollector` to use `Path::join()` at the same time, though I won't block merging this PR for those.

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (GuySartorelli) [reply src/i18n/TextCollection/i18nTextCollector.php]:
```suggestion
                return array_key_exists($name, $allModules) ? $this->getModuleName($name, $module) : null;
```

```php
private function getModuleName(string $origName, Module $module)
{
    return strpos($origName, self::THEME_PREFIX) === 0 ? $origName : $module->getName();
}
```

(see https://github.com/silverstripe/silverstripe-framework/pull/10685#discussion_r1187006125 for the constant)

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (GuySartorelli) [inline src/i18n/TextCollection/i18nTextCollector.php]:
```suggestion
            $moduleName = $this->getModuleName($moduleName, $module);
```
See https://github.com/silverstripe/silverstripe-framework/pull/10685#discussion_r1187010570 for the `getModuleName()` method

**Review by GuySartorelli — DISMISSED** [review]
Please also add some unit tests to validate that this new functionality actually collects translations from themes as expected.

Sorry there are so many changes I didn't mention last time - I should have caught these the first time around.

**Comment by GuySartorelli** [comment]
I've created https://github.com/silverstripe/developer-docs/pull/263 which will document this change.

File: src/i18n/Messages/YamlWriter.php
Comment (xini) [reply src/i18n/Messages/YamlWriter.php]: done.

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php:470]: went with option 1. done.

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php]: done.

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php]: done.

File: src/i18n/TextCollection/i18nTextCollector.php
Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php]: where do I get the `$module` parameter from? Sprry, too many nested methods for bmy brain...

Comment (GuySartorelli) [reply src/i18n/TextCollection/i18nTextCollector.php]: The `getModulesAndThemes()` method returns an array of `$name: string => $module: Module` doesn't it? So `$allModules[$name]`

Comment (xini) [reply src/i18n/TextCollection/i18nTextCollector.php]: Yes, correct. Thanks!

**Comment by xini** [comment]
Hey @GuySartorelli, I think this is now all cleaned up I have added a test as well. Let me know if you're happy with this and I'll squash the commits. Thank you!

**Comment by GuySartorelli** [comment]
That's awesome, thank you! I haven't run it locally yet which I'll do before merging, but given the test is passing I'm pretty confident it works. The code looks good from a visual-inspection. Please go ahead and squash it, I'll try to make time either later today or tomorrow to run it locally before merging.

**Comment by xini** [comment]
Done, thank you!

**Review by GuySartorelli — APPROVED** [review]
Works great locally, thank you! Between this and the ORM change, this has been a significant contribution.

**Comment by xini** [comment]
Thank you! It has been bugging me since 4.0, when the themes functionality was removed... 🙃 And with a few non-english projects over the last few months I finally had enough... 😉

**Pull request merged**
